### ISSUE TITLE
Need to be able to use custom HK2 ServiceLocator for Spring integration

### ISSUE THREAD

**Issue by jacek99** [dropwizard/dropwizard#1736]
We have been one of the early adopters of DW and been using it in production for over 3 years.
However, we are stuck on 0.7 and cannot move to a higher version.

The reason for this is the *******\* HK2 DI that Jersey 2 introduced (and DW adopted in 0.8). It just refuses to play nice with the our Spring context that we use everywhere for DI.

We've had 2 senior architects spend 3 weeks trying to get it to work and HK2 fails to play nicely with Spring no matter what. It wants to @Inject its own beans into our JAX-RS classes, even though those are existing Spring beans and have already been injected with existing Spring beans.

The only solution seems to write a custom HK2 service locator that just fetches the beans directly from Spring and hope that works.

However, Dropwizard seems to offer no API to allow defining a custom ServiceLocator in HK2 to override the default one.

If we cannot get this to work, we cannot be stuck on 0.7 and will be faced with the difficult choice of moving to a different platform...which we would hate, since we all love Dropwizard.

Help please :-/

**Comment by nickbabcock** [comment]
Thank you for being such a long time user of Dropwizard! Stories of such usage is very heartwarming.

Forewarning, I'm not expert in HK2 or Spring, but it looks like the [dropwizard-guice](https://github.com/HubSpot/dropwizard-guice) project accomplishes what you want (ob. for guice not Spring). I'd recommend examining the project. [These lines](https://github.com/HubSpot/dropwizard-guice/blob/2f207a23d8975436d2a48f8ac720c4befeb4753d/src/main/java/com/hubspot/dropwizard/guice/GuiceBundle.java#L102-L112) and [these lines](https://github.com/Squarespace/jersey2-guice/blob/08d3e30a89dc3cd27f466391ea17cd404a4d225c/jersey2-guice-impl/src/main/java/com/squarespace/jersey2/guice/JerseyGuiceUtils.java#L235-L277) should be a good start 😃

**Comment by jacek99** [comment]
I had a look at it and frankly speaking, it's not very clear how he accomplishes it (since I am not a Guice expert either).

From what I read on HK2 at the core of its is the ServiceLocator object. It seems that if we were able to implement our own version of that interface and just configure HK2 within Jersey to use it (instead of the one created by default by Dropwizard), that *should* do the trick.

However, there seems to be no way to customize this within the current DW API....

**Comment by jacek99** [comment]
Basically, all I need to tell that bloody HK2 (which I curse Sun every day for introducing it and ruining Jersey for everybody) is to fetch an existing Spring bean from our existing context instead of trying to create it itself and inject it.

**Comment by janolaveide** [comment]
In DW 0.8 I am using Spring to @Inject into my jax-rs classes without
problems. I have written a Spring bundle that does most of the heavy
lifting. HK2 does not get in the way. I am planning a blog post on this
soon. Feel free to ask if you have any specific questions.

Den torsdag 22. september 2016 skrev Jacek Furmankiewicz <
notifications@github.com> følgende:

**Comment by jacek99** [comment]
We have the opposite. All of our classes (including JAX-RS resources) were Spring beans from the beginning and we @Inject other Spring beans (e.g. DAOs) into them.

Where the bloody HK2 gets in the way is that now wants to also implement the @Inject itself (even though Spring has already done it), so HK2 gives an an error like "unable to find bean of type SomeEntityDao" (as that bean is a Spring bean, not an HK2 bean).

Basically, HK2 seems to want to manage JAX-RS resources by itself and does not handle the case at all where it's already part of another DI engine context.

This worked perfectly fine and was working great prior to Jersey 2, we had a DW app and Spring to manage all the beans inside it. Then came Jersey 2 with HK2 and ruined everything.

**Comment by janolaveide** [comment]
All my resources are managed by Spring through a @Component annotation an
package scanning. They have services injected which themselves have daos
injected etc. Works fine, never had hk2 getting in the way

Den torsdag 22. september 2016 skrev Jacek Furmankiewicz <
notifications@github.com> følgende:

**Comment by jacek99** [comment]
Are your JAX-RS resources Spring beans as well? When you have @Inject on those, who does the injection? HK2 or Spring?

Every time HK2 sees an @Inject in a JAX-RS resource, it wants to inject that resource itself. We can't stop it from doing that.

Simple example;

``
@Service    // Spring Service bean itself
@Path("/mealService")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_FORM_URLENCODED)
public class MealServiceResource {

     @Inject
     private SomeSpringDao dao;

``

**Comment by janolaveide** [comment]
Yes they are, Spring does the injection. Have you tried constructor
injection? I am not using field injection the way you do, otherwise my
resources  are similar to your example

Den torsdag 22. september 2016 skrev Jacek Furmankiewicz <
notifications@github.com> følgende:

**Comment by jacek99** [comment]
So how do you stop HK2 from trying to @Inject inject that SomeSpringDao bean?

Switching to contructor injection would be ugly and require rewriting hundreds of JAX-RS resources across multiple apps that use the same base framework....

**Comment by janolaveide** [comment]
I don't stop it. It never tries to. 

Sendt fra min iPad

**Comment by jacek99** [comment]
Ah, sorry, just read your comment. So you do not use @Inject for Spring beans on JAX-RS resources, versus we do.

So your solution would not work for us.

**Comment by janolaveide** [comment]
Yes I do
@component on jax-rs resource, constructor injection and getting beans with
annotation @path from app context and registrering these with dw

Den torsdag 22. september 2016 skrev Jacek Furmankiewicz <
notifications@github.com> følgende:

**Comment by jacek99** [comment]
Well, I am glad it works, but we do not use constructor injection anywhere in our, we use @Inject everywhere for consistency (and the code has much less boilerplate).

A cleaner IMHO solution would still be to drive HK2 with a pass-through ServiceLocator that just gives it any bean it wants from an already existing Spring context.

But we need DW to give us the ability to override the ServiveLocator within Jersey, which right now the API does not allow us to.

**Comment by janolaveide** [comment]
Just to be clear, i do use  @inject on the constructor, but if you cannot
change that i guess you are out of options  for the time being. spring boot
next?

Den torsdag 22. september 2016 skrev Jacek Furmankiewicz <
notifications@github.com> følgende:

**Comment by jacek99** [comment]
Hence, the initial request in this ticket.

HK2 allows you to use your own ServiceLocator, it's part of its design:

https://hk2.java.net/2.5.0-b07/apidocs/org/glassfish/hk2/api/ServiceLocator.html

They provide a default implementation:
https://hk2.java.net/2.5.0-b16/getting-started.html

but you are supposed to be able to use your own.

Unfortunately, Dropwizard does not allow us to use our own, as it initializes Jersey and HK2 without any hooks to allow injecting your own ServiceLocator.

That is what we would love to get fixed.

**Comment by nickbabcock** [comment]
Have you tried looking at [Spring/HK2 Bridge](https://hk2.java.net/spring-bridge/), like how [dropwizard-guicey](https://github.com/xvik/dropwizard-guicey) uses [guice-bridge](https://hk2.java.net/guice-bridge/) in their [GuiceFeature](https://github.com/xvik/dropwizard-guicey/blob/54ca4bac16a44d0a5de4c6828f2202629b3246fd/src/main/java/ru/vyarus/dropwizard/guice/module/jersey/GuiceFeature.java) to modify the service locator? It certainly looks like one can inject their own service locator.

**Comment by jacek99** [comment]
Hm we tried it before, a few months ago, but it did not seem to work for us.
According to the docs, it was supposed to just pick it up if it was in the classpath.

The docs you are pointing to seem to have some extra instructions on how to wire the two together.
Let me try that and see if that works. Maybe that could do it.

Will try and let you know if we hit any snags or unknowns. HK2 is somewhat of an undocumented beast.

**Comment by janolaveide** [comment]
I see the same behaviour as you describe with field injection, However, if
I change @Inject to @Autowired,  field injection also works without H2K
getting in the way. Maybe an alternative if the bridge approach does not
work.


On 22 September 2016 at 23:02, Jacek Furmankiewicz <notifications@github.com

**Comment by nickbabcock** [comment]
Has this been resolved?

**Comment by jacek99** [comment]
Sorry, have not had a chance to look at it again, other work-related issues swamped me. 
Hope to have another go at a 1.0 attempt within a month

**Comment by jacek99** [comment]
Hi Nick, unfortunately we cannot get the Spring/HK2 bridge to work.

The reason for that is that the ServiceLocator does not seem to be initialized yet when the run() is executed.

Here is my code in the run() that is executed after our Spring context starts:

```
        ServletContainer sc = (ServletContainer) environment.getJerseyServletContainer();
        ApplicationHandler ah = sc.getApplicationHandler();
        ServiceLocator hk2 = ah.getServiceLocator();
```

It fails with a NullPointerException on

```
        ApplicationHandler ah = sc.getApplicationHandler();
```

when I debug the code I see in Dropwiizard"

```
    @Override
    public ApplicationHandler getApplicationHandler() {
        return webComponent.appHandler;
    }
```

but unfortunately...webComponent = null at this point so it throws an NPE.

Any suggestions are welcome...how can I get a valid instance of the HK2 ServiceLocator at startup.

**Comment by jacek99** [comment]
Finally got it working, but it was a trip. Probably worth documenting in a Wiki or the webpage, so that everyone does not have to go through this obscure discovery process.

First, we need to define a Jersey Feature that is also a ServiceLocator provider and in there we enable the Spring/HK2 bridge on the alread created service locator:

```
import com.google.common.base.Preconditions;
import org.glassfish.hk2.api.ServiceLocator;
import org.glassfish.jersey.ServiceLocatorProvider;
import org.jvnet.hk2.internal.ServiceLocatorImpl;
import org.jvnet.hk2.spring.bridge.api.SpringBridge;
import org.jvnet.hk2.spring.bridge.api.SpringIntoHK2Bridge;
import org.springframework.context.annotation.AnnotationConfigApplicationContext;

import javax.inject.Provider;
import javax.ws.rs.core.Feature;
import javax.ws.rs.core.FeatureContext;

/**
 * Integrates the default Jersey HK2 ServiceLocator with Spring via the
 * Spring/HK2 bridge from Oracle
 *
 * @author Jacek Furmankewicz
 */
public class SpringAwareHK2ServiceLocatorFeature implements Feature, Provider<ServiceLocator> {

    private AnnotationConfigApplicationContext springContext;
    private ServiceLocator locator;

    public SpringAwareHK2ServiceLocatorFeature(AnnotationConfigApplicationContext springContext) {
        this.springContext = springContext;
    }

    @Override
    public ServiceLocator get() {
        return Preconditions.checkNotNull(locator, "Service locator is not yet available");
    }

    @Override
    public boolean configure(FeatureContext context) {
        locator =  ServiceLocatorProvider.getServiceLocator(context);

        SpringBridge.getSpringBridge().initializeSpringBridge(locator);
        SpringIntoHK2Bridge springBridge = locator.getService(SpringIntoHK2Bridge.class);
        springBridge.bridgeSpringBeanFactory(springContext.getBeanFactory());

        return true;
    }
}
```

Then we need to register it manually at startup, already initialzed with an existing Spring context:

```
        // hook up Spring context and Jersey HK2 together
        environment.jersey().register(new SpringAwareHK2ServiceLocatorFeature(context));
```

That finally did it, without any further modifications to Dropwizard.

Now Jersey/HK2 properly gets Spring beans and does not complain that they don't exist.

Thanks for all your help. The key code was in the Guice / Dropwizard module and I adapted it from there for Spring.

Once again, I think it's worthwhile adding a  "Dropwizard & Spring" section on the main site and documenting this there.

**Comment by joschi** [comment]
@jacek99 Thank you so much for documenting this!

We'd love to include this into the documentation. Do you have time to create a small PR for it?

**Comment by jacek99** [comment]
Sure, just point you to where i could do this.

**Comment by joschi** [comment]
@jacek99 It would be great if you could include a paragraph in the dropwizard-core documentation at https://github.com/dropwizard/dropwizard/blob/master/docs/source/manual/core.rst

**Comment by jacek99** [comment]
Sounds good. Will try to do this one of the evenings this week.

**Comment by jacek99** [comment]
Any particular section you think it belongs to?

Or just add it to the end maybe as a separate section like

* Integration with 3rd party libraries
** Spring

etc

**Comment by joschi** [comment]
@jacek99 Yes, please just add a separate section at the end of the file.

**Comment by rvowles** [comment]
The reason @Component and @Autowired work is that the Spring/Jersey library specifically looks for those and thunks them in. If you actually use Spring without classpath scanning or @Component or old-school auto-wired, you have to provide your own service locator.

**Comment by nickbabcock** [comment]
Closing this issue as it seems to have gone stale. We'd love to see Spring interop documented but we're not avid spring users so the community will need to step up. Jersey 2.26 may have changed the interop scenario as well. If discussion continues on this issue, we're happy to open it back up.
